In [1]:
%cd "C:\Users\loren\Documents\COMP3608-Project"
import pandas as pd

#Algorithms used
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import LinearSVC

#Scalers to test
from sklearn.discriminant_analysis import StandardScaler
from sklearn.preprocessing import MinMaxScaler

#Sampling techniques to test
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import RandomUnderSampler
from imblearn.over_sampling import RandomOverSampler

from DataSet3.HelpFunctions.comparison_pipline import model_comparison

C:\Users\loren\Documents\COMP3608-Project


c:\Python312\Lib\site-packages\IPython\core\magics\osm.py:417: UserWarning: This is now an optional IPython functionality, setting dhist requires you to install the `pickleshare` library.
  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]


In [2]:
#Model, resampling techniques and sampling methods to be compared are all added to a dict
models_to_test = {
    'Logistic Regression': LogisticRegression(random_state=42, max_iter=1000, n_jobs=-1),
    'Random Forest': RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1),
    'Linear SVC': LinearSVC(dual=False, random_state=42, max_iter=2000)
}

scalers = {
        'None': None,
        'StandardScaler': StandardScaler(),
        'MinMaxScaler': MinMaxScaler()
}

samplers = {
    'None': None,
    'SMOTE': SMOTE(random_state=42), # Keep SMOTE
    'RUS': RandomUnderSampler(random_state=42), # Add RUS
    'ROS': RandomOverSampler(random_state=42)   # Add ROS
}

In [3]:

df = pd.read_csv('card_transdata.csv', index_col=0)
binary_columns = {'repeat_retailer': 'bool', 'used_chip': 'bool',
                  'used_pin_number': 'bool', 'online_order': 'bool', 'fraud': 'bool'}
df = df.astype(binary_columns)


In [4]:

# Define features and target
X = df.drop(columns=["fraud"])
y = df["fraud"]

numerical_cols_actual = [col for col in X.columns if X[col].dtype in ['float64', 'int64', 'float32', 'int32']]
print(f"Using numerical columns for scaling: {numerical_cols_actual}")


Using numerical columns for scaling: ['distance_from_last_transaction', 'ratio_to_median_purchase_price']


In [5]:
# Run the pipeline
all_results = model_comparison(models_to_test, scalers, samplers, X, y, numerical_cols_actual, verbose=1)

Splitting data: test_size=0.2, random_state=42, stratify=y
Train shape: (800000, 6), Test shape: (200000, 6)
Test label distribution:
fraud
False    0.912595
True     0.087405
Name: proportion, dtype: float64


=== Evaluating Model: Logistic Regression ===

=== Evaluating Model: Random Forest ===

=== Evaluating Model: Linear SVC ===
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for ROC AUC (Linear SVC).
Info: Used decision_function() for RO

In [6]:
all_results

Model          Scaler Sampler  Accuracy  Precision  \
0         Random Forest    MinMaxScaler    None  0.981155   0.997749   
1         Random Forest            None    None  0.980790   0.992347   
2         Random Forest  StandardScaler    None  0.980780   0.991849   
3         Random Forest    MinMaxScaler     ROS  0.978430   0.957156   
4         Random Forest            None     ROS  0.976125   0.925976   
5         Random Forest  StandardScaler     ROS  0.976010   0.924607   
6         Random Forest    MinMaxScaler   SMOTE  0.952125   0.694155   
7         Random Forest            None   SMOTE  0.931895   0.577604   
8         Random Forest  StandardScaler   SMOTE  0.931165   0.574144   
9   Logistic Regression            None    None  0.954775   0.921042   
10  Logistic Regression  StandardScaler    None  0.954755   0.920927   
11        Random Forest    MinMaxScaler     RUS  0.913450   0.502949   
12        Random Forest  StandardScaler     RUS  0.901355   0.464576   
13        Random Forest            None     RUS  0.901045   0.463627   
14           Linear SVC  StandardScaler     ROS  0.886700   0.423494   
15           Linear SVC            None     ROS  0.886695   0.423481   
16           Linear SVC  StandardScaler   SMOTE  0.886530   0.423046   
17           Linear SVC            None   SMOTE  0.886455   0.422768   
18           Linear SVC  StandardScaler     RUS  0.886045   0.421755   
19           Linear SVC            None     RUS  0.886045   0.421751   
20           Linear SVC    MinMaxScaler     ROS  0.885770   0.421063   
21           Linear SVC    MinMaxScaler   SMOTE  0.885620   0.420701   
22  Logistic Regression            None     RUS  0.882435   0.413735   
23  Logistic Regression  StandardScaler     RUS  0.882400   0.413657   
24  Logistic Regression            None     ROS  0.881915   0.412558   
25  Logistic Regression  StandardScaler     ROS  0.881850   0.412400   
26  Logistic Regression  StandardScaler   SMOTE  0.881790   0.412274   
27  Logistic Regression            None   SMOTE  0.881775   0.412159   
28           Linear SVC    MinMaxScaler     RUS  0.878735   0.403665   
29  Logistic Regression    MinMaxScaler    None  0.942765   0.926733   
30  Logistic Regression    MinMaxScaler     ROS  0.875065   0.395617   
31  Logistic Regression    MinMaxScaler   SMOTE  0.874865   0.395194   
32           Linear SVC  StandardScaler    None  0.938895   0.911066   
33           Linear SVC            None    None  0.938895   0.911066   
34           Linear SVC    MinMaxScaler    None  0.938065   0.909749   
35  Logistic Regression    MinMaxScaler     RUS  0.842600   0.335372   

      Recall  F1 Score   ROC AUC  
0   0.786168  0.879411  0.958454  
1   0.786282  0.877378  0.946634  
2   0.786568  0.877361  0.945886  
3   0.788513  0.864689  0.958992  
4   0.790001  0.852601  0.947014  
5   0.789943  0.851987  0.947308  
6   0.808478  0.746968  0.962865  
7   0.821749  0.678378  0.961545  
8   0.822607  0.676276  0.962301  
9   0.527830  0.671079  0.928591  
10  0.527659  0.670910  0.928608  
11  0.834048  0.627502  0.963127  
12  0.843258  0.599094  0.963100  
13  0.842172  0.598030  0.962914  
14  0.819976  0.558526  0.937437  
15  0.819976  0.558515  0.937440  
16  0.819690  0.558070  0.937440  
17  0.818546  0.557562  0.937068  
18  0.818660  0.556707  0.937122  
19  0.818603  0.556690  0.937123  
20  0.818546  0.556078  0.937128  
21  0.818660  0.555789  0.937150  
22  0.827470  0.551647  0.938846  
23  0.827527  0.551590  0.938859  
24  0.828042  0.550726  0.938884  
25  0.827985  0.550572  0.938878  
26  0.828156  0.550498  0.938881  
27  0.827241  0.550193  0.938625  
28  0.811624  0.539170  0.934627  
29  0.374807  0.533746  0.918991  
30  0.813683  0.532385  0.934880  
31  0.813855  0.532039  0.934895  
32  0.333448  0.488211  0.924636  
33  0.333448  0.488211  0.924636  
34  0.323494  0.477276  0.924119  
35  0.815686  0.475316  0.924706

based on the results, `random forest` with no sampling techniques and `MinMaxScaler` gave the best f1-score of 0.88 and on of the best ROC AUC. The model with the best ROC AUC is `LinearSVC` with `ROS` (random under sampling) and a `StandardScaler` with a value of 0.937 (row 14 in the results table).

Random Forest	MinMaxScaler	None	0.981155	0.997749	0.786168	0.879411	0.958454
14	Linear SVC	StandardScaler	ROS	0.886700	0.423494	0.819976	0.558526	0.937437

## How Sampling technique works
add mathematical models and explain each one

## How Scaling Works
add mathematical models and explain each one